### PR TITLE
PHOENIX-7040 : Support TTL for views using the new column TTL in SYSTEM.CATALOG

### PR DESCRIPTION
**Pull request by lokiore**
- Enabled TTL for Updatable Views
- Store TTL at only one level where TTL is defined in hierarchy.
- getTTL() should get TTL from hierarchy so while building PTable store get TTL from hierarchy and store it in PTable.
- Altering should be allowed if TTL is not defined in up/down the hierarchy before.

Altering TTL of a parent is not reflected directed to it's child's getTTL(), need to work on invalidate Cache for this.

### CODE REVIEW

**Comment by lokiore** [comment]
I'll clean up the PR after tests are run.
Need to add more tests as well

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/TTLAsPhoenixTTLIT.java [hunk 1]
```diff
@@ -370,4 +455,62 @@ private String createViewOnViewWithTTL(Connection conn, String parentViewName,
         return childView;
     }
 
+
```
Comment (jpisaac) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/TTLAsPhoenixTTLIT.java:458]: You may need to add tests for resetting TTL (TTL=NONE) at a given level and then setting (ALTER) to new TTL at a different level.

Currently Alter table  TTL=NONE sets it to 0 and thus cannot set a new TTL at a different level.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java [hunk 2]
```diff
@@ -63,6 +63,7 @@
 import java.util.Map;
 import java.util.Properties;
 
+import org.apache.commons.lang3.mutable.MutableBoolean;
```
Comment (jpisaac) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java:66]: nit: Unused imports

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewUtilIT.java [hunk 3]
```diff
@@ -17,6 +17,7 @@
  */
 package org.apache.phoenix.end2end;
 
+import org.apache.commons.lang3.mutable.MutableBoolean;
```
Comment (jpisaac) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewUtilIT.java:20]: nit: Unused imports

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java [hunk 4]
```diff
@@ -19,78 +19,20 @@
...
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.TTL_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.TYPE_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.UPDATE_CACHE_FREQUENCY_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.USE_STATS_FOR_PARALLELIZATION_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.VIEW_CONSTANT_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.VIEW_INDEX_ID_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.VIEW_INDEX_ID_DATA_TYPE_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.VIEW_STATEMENT_BYTES;
-import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.VIEW_TYPE_BYTES;
+import static org.apache.phoenix.jdbc.PhoenixDatabaseMetaData.*;
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:22]: nit: removed .* imports

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java [hunk 5]
```diff
@@ -629,6 +542,11 @@ public void start(CoprocessorEnvironment env) throws IOException {
         Tracing.addTraceMetricsSource();
         Metrics.ensureConfigured();
         metricsSource = MetricsMetadataSourceFactory.getMetadataMetricsSource();
+
+        //should I load it here or everytime in get TableFromCells.
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:546]: Getting a Table is a light weight operation. See conn.getTable() help
"This is a lightweight operation, pooling or caching of the returned Table is neither required nor desired."

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java [hunk 6]
```diff
@@ -681,6 +599,13 @@ public void getTable(RpcController controller, GetTableRequest request,
                     table = ViewUtil.addDerivedColumnsFromParent(connection, table, pTable);
                 }
             }
+//
+//            if (table.getType() == PTableType.INDEX) {
+//                try (PhoenixConnection connection = QueryUtil.getConnectionOnServer(env.getConfiguration()).unwrap(PhoenixConnection.class)) {
+//                    PTable pTable = PhoenixRuntime.getTableNoCache(connection, table.getParentName().getString());
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:605]: nit : remove commented code

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java [hunk 7]
```diff
@@ -1400,12 +1325,23 @@ private PTable getTableFromCells(List<Cell> tableCellList, List<List<Cell>> allC
...
                         ttlKv.getValueOffset(), SortOrder.getDefault());
-        builder.setTTL(ttlKv != null ? ttl : oldTable != null
-                ? oldTable.getTTL() : TTL_NOT_DEFINED);
+        ttl = ttlKv != null ? ttl : oldTable != null
+                ? oldTable.getTTL() : TTL_NOT_DEFINED;
+        if (tableType == VIEW && ttl == TTL_NOT_DEFINED) {
+            //Scan SysCat to get TTL from Parent View/Table
+            byte[] viewKey = SchemaUtil.getTableKey(tenantId == null ? null : tenantId.getBytes(),
+                    schemaName == null ? null : schemaName.getBytes(), tableNameBytes);
+            dumpTable(sysCat);
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:1339]: nit: add TODO to remove dumpTable

**Review by jpisaac — COMMENTED** [review]
Took a first pass, Will do some more testing,

**Comment by jpisaac** [comment]
Setting TTL=NONE at the child level and then try to set to some value at a global view level gives the following exception.
Inspecting the TTL value after setting it to NONE, shows a value of "0".

Caused by: org.apache.hadoop.hbase.ipc.RemoteWithExtrasException(org.apache.hadoop.hbase.DoNotRetryIOException): org.apache.hadoop.hbase.DoNotRetryIOException: ERROR 10956 (44A37): ERROR 10956 (44A37): TTL property is already defined in hierarchy for this entity tableName=[84, 69, 83, 84, 95, 69, 78, 84, 73, 84, 89].[71, 49, 95, 49, 48, 49] TEST_ENTITY.G1_101
	at org.apache.phoenix.util.ServerUtil.createIOException(ServerUtil.java:109)
	at org.apache.phoenix.util.ServerUtil.throwIOException(ServerUtil.java:83)
	at org.apache.phoenix.coprocessor.MetaDataEndpointImpl.mutateColumn(MetaDataEndpointImpl.java:3555)
	at org.apache.phoenix.coprocessor.MetaDataEndpointImpl.addColumn(MetaDataEndpointImpl.java:3641)
	at org.apache.phoenix.coprocessor.generated.MetaDataProtos$MetaDataService.callMethod(MetaDataProtos.java:17569)
	at org.apache.hadoop.hbase.regionserver.HRegion.execService(HRegion.java:7936)
	at org.apache.hadoop.hbase.regionserver.RSRpcServices.execServiceOnRegion(RSRpcServices.java:2535)
	at org.apache.hadoop.hbase.regionserver.RSRpcServices.execService(RSRpcServices.java:2509)
	at org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos$ClientService$2.callBlockingMethod(ClientProtos.java:45957)

File: phoenix-core/src/main/java/org/apache/phoenix/util/ViewUtil.java [hunk 8]
```diff
@@ -132,20 +134,35 @@ public static Pair<List<PTable>, List<TableInfo>> findAllDescendantViews(
...
+           List<TableInfo> orphanChildViews, boolean findJustOneLegitimateChildView)
+           throws IOException, SQLException{
+
+         return findAllDescendantViews(sysCatOrsysChildLink, null, serverSideConfig,
+                 parentTenantId, parentSchemaName, parentTableOrViewName, clientTimeStamp,
+                 legitimateChildViews, orphanChildViews, findJustOneLegitimateChildView,
+                 new Pair<>(false, false));
+
+    }
+
```
Comment (jpisaac) [inline phoenix-core/src/main/java/org/apache/phoenix/util/ViewUtil.java:154]: nit: Can u add some comments for the new methods and hierarchy traversal

**Review by jpisaac — APPROVED** [review]
Thanks @lokiore
